### PR TITLE
Extend Recover API so that we can update media source for items

### PR DESCRIPTION
**Pull request by sangwoo108**
The media source url could be easily expired. So we might have got to update media source url before we try recovering the media file. In case we can notice that the url would be expired from url, we can update media source first. But mostly we can't infer if its expired or not. So when we fail to recover media file, we try updating media source by default.

In order to implement this new behavior, also refactored PlaylistMediaFileDownloadManager. The delegate pattern is extracted out to callbacks so that we can remember |update_media_src_and_retry_on_failure| param,

Resolves https://github.com/brave/brave-browser/issues/28291

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Automated
PlaylistServiceUnitTest.MediaRecoverTest

### CODE REVIEW

File: components/playlist/browser/playlist_service.cc [hunk 1]
```diff
@@ -747,6 +798,41 @@ void PlaylistService::RemoveLocalDataForItem(
...
+  DCHECK(IsValidPlaylistItem(item->id));
+
+  VLOG(2) << __func__ << ": " << item->id << " result path" << media_file_path;
+
+  if (media_file_path.empty() && update_media_src_and_retry_on_fail) {
+    VLOG(2) << __func__ << ": " << item->id
+            << " try updating media src and download";
+
+    base::SequencedTaskRunner::GetCurrentDefault()->PostTask(
+        FROM_HERE, base::BindOnce(&PlaylistService::RecoverLocalDataForItem,
```
Comment (simonhong) [inline components/playlist/browser/playlist_service.cc:815]: Is there any possibility that "download failed and recover" will be repeated indefinitely?

Comment (sangwoo108) [reply components/playlist/browser/playlist_service.cc:815]: Well spotted! I should count the number of retrial.

Comment (sangwoo108) [reply components/playlist/browser/playlist_service.cc:815]: Oh, it's not likely happen, as update_media_src_and_retry_on_fail will be false when update_media_src_before_recovery is true.

File: components/playlist/browser/playlist_service.cc [hunk 2]
```diff
@@ -747,6 +798,41 @@ void PlaylistService::RemoveLocalDataForItem(
...
+    mojom::PlaylistItemPtr item,
+    const std::string& media_file_path) {
+  DCHECK(item);
+  DCHECK(IsValidPlaylistItem(item->id));
+
+  VLOG(2) << __func__ << ": " << item->id << " result path" << media_file_path;
+
+  if (media_file_path.empty() && update_media_src_and_retry_on_fail) {
+    VLOG(2) << __func__ << ": " << item->id
+            << " try updating media src and download";
```
[the remark this answers is not in the document]
Comment (sangwoo108) [reply components/playlist/browser/playlist_service.cc]: Yup. Done.

File: components/playlist/browser/playlist_service.cc (hunk 1 above)
Comment (simonhong) [reply components/playlist/browser/playlist_service.cc:820]: Ok, great. Thanks for checking!

**Review by simonhong — APPROVED** [review]
++

**Pull request merged**